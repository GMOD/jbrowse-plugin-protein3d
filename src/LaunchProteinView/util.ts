import { Feature } from '@jbrowse/core/util'
import { ungzip } from 'pako'
import { useEffect, useState } from 'react'

async function myfetch(url: string) {
  const res = await fetch(url)
  if (!res.ok) {
    throw new Error(`HTTP ${res.status} fetching ${url}: ${await res.text()}`)
  }
  return res.arrayBuffer()
}

export interface Row {
  gene_id: string
  gene_id_version: string
  transcript_id_version: string
  transcript_id: string
  pdb_id: string
  refseq_mrna_predicted_id: string
  refseq_mrna_id: string
}
export function useBiomartMappings(url: string) {
  const [data, setData] = useState<Row[]>()
  const [error, setError] = useState<unknown>()
  useEffect(() => {
    ;(async () => {
      try {
        const d = new TextDecoder('utf8')
          .decode(ungzip(await myfetch(url)))
          .split('\n')
          .slice(1)
          .filter((f) => !!f)
          .map((f) => {
            const res = f.split('\t')
            const [
              gene_id,
              gene_id_version,
              transcript_id,
              transcript_id_version,
              pdb_id,
              refseq_mrna_id,
              refseq_mrna_predicted_id,
            ] = res
            return {
              gene_id,
              gene_id_version,
              transcript_id_version,
              transcript_id,
              pdb_id,
              refseq_mrna_predicted_id,
              refseq_mrna_id,
            }
          })
        setData(d)
      } catch (e) {
        setError(e)
      }
    })()
  }, [url])
  return [data, error] as const
}

export function check(row: Row, val: string) {
  return (
    (row.transcript_id === val ||
      row.refseq_mrna_id === val ||
      row.transcript_id_version === val) &&
    row.pdb_id
  )
}

export function getTranscriptFeatures(feature: Feature) {
  // check if we are looking at a 'two-level' or 'three-level' feature by
  // finding exon/CDS subfeatures. we want to select from transcript names
  const subfeatures = feature.get('subfeatures') || []
  return subfeatures.some(
    (f) => f.get('type') === 'CDS' || f.get('type') === 'exon',
  )
    ? [feature]
    : subfeatures
}
