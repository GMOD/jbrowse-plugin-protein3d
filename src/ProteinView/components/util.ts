export async function loadStructure({
  url,
  file,
  plugin,
}: {
  url?: string
  file?: { type: string; filestring: string }
  plugin?: {
    clear: () => void
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any
  }
}) {
  if (!plugin) {
    return
  }
  plugin.clear()
  console.log('wow clear')
  if (file) {
    const { filestring, type } = file
    const data = await plugin.builders.data.rawData({
      data: filestring,
    })
    const traj = await plugin.builders.structure.parseTrajectory(data, type)
    await plugin.builders.structure.hierarchy.applyPreset(traj, 'default')
  } else {
    const structureUrl = url
    if (!structureUrl) {
      return
    }
    const data = await plugin.builders.data.download(
      { url: structureUrl },
      { state: { isGhost: true } },
    )
    let ext = structureUrl.split('.').pop()?.replace('cif', 'mmcif')
    if (ext?.includes('?')) {
      ext = ext.slice(0, Math.max(0, ext.indexOf('?')))
    }
    console.log({ ext, data })
    const traj = await plugin.builders.structure.parseTrajectory(data, ext)
    await plugin.builders.structure.hierarchy.applyPreset(traj, 'default')
  }
}
