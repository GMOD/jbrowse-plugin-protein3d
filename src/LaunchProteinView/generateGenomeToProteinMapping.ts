import { Feature, notEmpty } from '@jbrowse/core/util'

export interface Alignment {
  alns: {
    id: string
    seq: string
  }[]
}

function align(alignment: Alignment, query: number) {
  const k1 = alignment.alns[0].seq
  const k2 = alignment.alns[1].seq
  if (k1.length !== k2.length) {
    throw new Error('mismatched length')
  }

  // let k1 =
  //   'MGQKGHKDSLYPCGGTPESSLHEALDQCMTALDLFLTNQFSEALSYLKPRTKESMYHSLTYATILEMQAMMTFDPQDILLAGNMMKEAQMLCQRHRRKSSVTDSFSSLVNRPTLGQFTEEEIHAEVCYAECLLQRAALTFLQGSSHGGAVRPRALHDPSHACSCPPGPGRQHLFLLQDENMVSFIKGGIKVRNSYQTYKELDSLVQSSQYCKGENHPHFEGGVKLGVGAFNLTLSMLPTRILRLLEFVGFSGNKDYGLLQLEEGASGHSFRSVLCVMLLLCYHTFLTFVLGTGNVNIEEAEKLLKPYLNRYPKGAIFLFFAGRIEVIKGNIDAAIRRFEECCEAQQHWKQFHHMCYWELMWCFTYKGQWKMSYFYADLLSKENCWSKATYIYMKAAYLSMFGKEDHKPFGDDEVELFRAVPGLKLKIAGKSLPTEKFAIRKSRRYFSSNPISLPVPALEMMYIWNGYAVIGKQPKLTDGILEIITKAEEMLEKGPENEYSVDDECLVKLLKGLCLKYLGRVQEAEENFRSISANEKKIKYDHYLIPNALLELALLLMEQDRNEEAIKLLESAKQNYKNYSMESRTHFRIQAATLQAKSSLENSSRSMVSSVSL-'
  // let k2 =
  //   '--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------MKAAYLSMFGKEDHKPFGDDEVELFRAVPGLKLKIAGKSLPTEKFAIRKSRRYFSSNPISLPVPALEMMYIWNGYAVIGKQPKLTDGILEIITKAEEMLEKGPENEYSVDDECLVKLLKGLCLKYLGRVQEAEENFRSISANEKKIKYDHYLIPNALLELALLLMEQDRNEEAIKLLESAKQNYKNYSMESRTHFRIQAATLQAKSSLENSSRSMVSSVSL*'
  let j = 0
  let k = 0

  // eslint-disable-next-line unicorn/no-for-loop
  for (let i = k; i < k1.length; i++) {
    const char1 = k1[i]
    const char2 = k2[i]

    if (char1 === char2) {
      k++
      j++
    } else if (char2 === '-') {
      j++
    } else if (char1 === '-') {
      k++
    }
    if (query === k) {
      return char2 === '-' ? undefined : j
    }
  }
  return undefined
}

// see similar function in msaview plugin
export function generateGenomeToProteinMapping({
  feature,
  alignment,
}: {
  feature: Feature
  alignment?: Alignment
}) {
  let iter = 0
  if (!alignment) {
    return []
  }
  const strand = feature.get('strand')
  const subs = feature.children() ?? []
  const cds = subs.filter(f => f.get('type') === 'CDS')
  if (strand === -1) {
    cds.sort((a, b) => b.get('start') - a.get('start'))
  } else {
    cds.sort((a, b) => a.get('start') - b.get('start'))
  }
  const s1 = alignment.alns[0].seq
  const s2 = alignment.alns[1].seq
  console.log({ s1, s2 })

  return cds
    .map(f => {
      const refName = f.get('refName').replace('chr', '')
      const featureStart = f.get('start')
      const featureEnd = f.get('end')
      const phase = f.get('phase')
      const len = featureEnd - featureStart
      const op = Math.floor(len / 3)
      const sourceProteinStart = iter
      const sourceProteinEnd = iter + op
      const targetProteinStart = align(alignment, sourceProteinStart)
      const targetProteinEnd = align(alignment, sourceProteinEnd)
      iter += op
      if (targetProteinStart !== undefined && targetProteinEnd !== undefined) {
        console.log({
          sourceProteinStart,
          sourceProteinEnd,
          targetProteinStart,
          targetProteinEnd,
          k1: s1.slice(sourceProteinStart, sourceProteinEnd),
          k2: s2.slice(targetProteinStart, targetProteinEnd),
        })
        return {
          refName,
          featureStart,
          featureEnd,
          sourceProteinStart,
          sourceProteinEnd,
          targetProteinStart,
          targetProteinEnd,
          phase,
          strand,
        } as const
      } else {
        return undefined
      }
    })
    .filter(notEmpty)
}
