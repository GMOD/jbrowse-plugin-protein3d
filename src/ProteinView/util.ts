import { Feature } from '@jbrowse/core/util'
import { Structure } from 'molstar/lib/mol-model/structure'
import { Script } from 'molstar/lib/mol-script/script'

export const proteinAbbreviationMapping = Object.fromEntries(
  [
    { name: 'alanine', abbreviation: 'Ala', singleLetterCode: 'A' },
    { name: 'arginine', abbreviation: 'Arg', singleLetterCode: 'R' },
    { name: 'asparagine', abbreviation: 'Asn', singleLetterCode: 'N' },
    { name: 'aspartic acid', abbreviation: 'Asp', singleLetterCode: 'D' },
    { name: 'cysteine', abbreviation: 'Cys', singleLetterCode: 'C' },
    { name: 'glutamic acid', abbreviation: 'Glu', singleLetterCode: 'E' },
    { name: 'glutamine', abbreviation: 'Gln', singleLetterCode: 'Q' },
    { name: 'glycine', abbreviation: 'Gly', singleLetterCode: 'G' },
    { name: 'histidine', abbreviation: 'His', singleLetterCode: 'H' },
    { name: 'isoleucine', abbreviation: 'Ile', singleLetterCode: 'I' },
    { name: 'leucine', abbreviation: 'Leu', singleLetterCode: 'L' },
    { name: 'lysine', abbreviation: 'Lys', singleLetterCode: 'K' },
    { name: 'methionine', abbreviation: 'Met', singleLetterCode: 'M' },
    { name: 'phenylalanine', abbreviation: 'Phe', singleLetterCode: 'F' },
    { name: 'proline', abbreviation: 'Pro', singleLetterCode: 'P' },
    { name: 'serine', abbreviation: 'Ser', singleLetterCode: 'S' },
    { name: 'threonine', abbreviation: 'Thr', singleLetterCode: 'T' },
    { name: 'tryptophan', abbreviation: 'Trp', singleLetterCode: 'W' },
    { name: 'tyrosine', abbreviation: 'Tyr', singleLetterCode: 'Y' },
    { name: 'valine', abbreviation: 'Val', singleLetterCode: 'V' },
  ].map(r => [r.abbreviation.toUpperCase(), r]),
)

export function checkHovered(hovered: unknown): hovered is {
  hoverFeature: Feature
  hoverPosition: { coord: number; refName: string }
} {
  return (
    !!hovered &&
    typeof hovered == 'object' &&
    'hoverFeature' in hovered &&
    'hoverPosition' in hovered
  )
}

export function getMolstarStructureSelection({
  structure,
  selectedResidue,
}: {
  structure: Structure
  selectedResidue: number
}) {
  return Script.getStructureSelection(
    Q =>
      Q.struct.generator.atomGroups({
        'residue-test': Q.core.rel.eq([
          Q.struct.atomProperty.macromolecular.label_seq_id(),
          selectedResidue,
        ]),
        'group-by': Q.struct.atomProperty.macromolecular.residueKey(),
      }),
    structure,
  )
}

export function toStr(r: {
  structureSeqPos: number
  code?: string
  chain?: string
}) {
  return [
    `Position: ${r.structureSeqPos}`,
    r.code
      ? `Letter: ${r.code} (${proteinAbbreviationMapping[r.code]?.singleLetterCode})`
      : '',
    r.chain ? `Chain: ${r.chain}` : '',
  ]
    .filter(f => !!f)
    .join(', ')
}

export function invertMap(arg: Record<number, number | undefined>) {
  return Object.fromEntries(Object.entries(arg).map(([a, b]) => [b, a]))
}
