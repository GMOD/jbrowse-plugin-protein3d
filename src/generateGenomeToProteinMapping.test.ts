import { SimpleFeature } from '@jbrowse/core/util'
import featureData from './data.json'
import { generateGenomeToProteinMapping } from './generateGenomeToProteinMapping'

test('mapping', () => {
  const feature = new SimpleFeature(featureData)
  const alignment = {
    consensus:
      '                                                                                                                                                                                                                                                                                                                          ...|:..|........|:.::...:..:....|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ',
    alns: [
      {
        id: 'a',
        seq: 'MAAAAQLSLTQLSSGNPVYEKYYRQVDTGNTGRVLASDAAAFLKKSGLPDLILGKIWDLADTDGKGILNKQEFFVALRLVACAQNGLEVSLSSLNLAVPPPRFHDTSSPLLISGTSAAELPWAVKPEDKAKYDAIFDSLSPVNGFLSGDKVKPVLLNSKLPVDILGRVWELSDIDHDGMLDRDEFAVAMFLVYCALEKEPVPMSLPPALVPPSKRKTWVVSPAEKAKYDEIFLKTDKDMDGFVSGLEVREIFLKTGLPSTLLAHIWSLCDTKDCGKLSKDQFALAFHLISQKLIKGIDPPHVLTPEMIPPSDRASLQKNIIGSSPVADFSAIKELDTLNNEIVDLQREKNNVEQDLKEKEDTIKQRTSEVQDLQDEVQRENTNLQKLQAQKQQVQELLDELDEQKAQLEEQLKEVRKKCAEEAQLISSLKAELTSQESQISTYEEELAKAREELSRLQQETAELEESVESGKAQLEPLQQHLQDSQQEISSMQMKLMEMKDLENHNSQLNWCSSPHSILVNGATDYCSLSTSSSETANLNEHVEGQSNLESEPIHQESPARSSPELLPSGVTDENEVTTAVTEKVCSELDNNRHSKEEDPFNVDSSSLTGPVADTNLDFFQSDPFVGSDPFKDDPFGKIDPFGGDPFKGSDPFASDCFFRQSTDPFATSSTDPFSAANNSSITSVETLKHNDPFAPGGTVVAASDSATDPFASVFGNESFGGGFADFSTLSKVNNEDPFRSATSSSVSNVVITKNVFEETSVKSEDEPPALPPKIGTPTRPCPLPPGKRSINKLDSPDPFKLNDPFQPFPGNDSPKEKDPEIFCDPFTSATTTTNKEADPSNFANFSAYPSEEDMIEWAKRESEREEEQRLARLNQQEQEDLELAIALSKSEISEA-',
      },
      {
        id: 'b',
        seq: '--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------MYLKSDSGLGGWITIPAVADVLKYSCIVCWSSREKNNVEQDLKEKEDTIKQRTSEVQDLQDEVQRENTNLQKLQAQKQQVQELLDELDEQKAQLEEQLKEVRKKCAEEAQLISSLKAELTSQESQISTYEEELAKAREELSRLQQETAELEESVESGKAQLEPLQQHLQDSQQEISSMQMKLMEMKDLENHNSQLNWCSSPHSILVNGATDYCSLSTSSSETANLNEHVEGQSNLESEPIHQESPARSSPELLPSGVTDENEVTTAVTEKVCSELDNNRHSKEEDPFNVDSSSLTGPVADTNLDFFQSDPFVGSDPFKDDPFGKIDPFGGDPFKGSDPFASDCFFRQSTDPFATSSTDPFSAANNSSITSVETLKHNDPFAPGGTVVAASDSATDPFASVFGNESFGGGFADFSTLSKVNNEDPFRSATSSSVSNVVITKNVFEETSVKSEDEPPALPPKIGTPTRPCPLPPGKRSINKLDSPDPFKLNDPFQPFPGNDSPKEKDPEIFCDPFTSATTTTNKEADPSNFANFSAYPSEEDMIEWAKRESEREEEQRLARLNQQEQEDLELAIALSKSEISEA*',
      },
    ],
  }
  expect(
    generateGenomeToProteinMapping({ feature, alignment }),
  ).toMatchSnapshot()
})
