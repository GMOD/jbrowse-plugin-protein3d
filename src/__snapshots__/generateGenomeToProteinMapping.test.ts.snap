// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping 1`] = `[]`;
